### ISSUE TITLE
Can not replicate PyTorch results using libtorch Backend of DD

### ISSUE THREAD

**Issue by YaYaB** [jolibrain/deepdetect#662]
#### Configuration
- Version of DeepDetect:
    - Locally compiled on Ubuntu 16.04 LTS
- Commit (shown by the server when starting):
    - Branch `Master`, 7512943bd830967c9c97f0780658d8e811c81c4d
- GPUS:
    - 1 x Nvidia GTX Titan X (Maxwell)
I used the torch and torchvision indicated with DD when it was compiled.

#### Your question / the problem you're facing:
I am not able to replicate results obtained using pytorch with DD and libtorch backend.
Let us take the model resnet50 as example (like the one showed in the integration PR).

The following script uses a resnet50 to compute prediction on an image :
Let us first create a repository and download a dog image :
```
mkdir replicate_pytorch_ddlibtorch
cd replicate_pytorch_ddlibtorch
mkdir data
cd data
wget https://www.learnopencv.com/wp-content/uploads/2019/05/dog.jpg
```

Then launch the following
```
import torch
from PIL import Image
from torchvision import transforms
import torchvision.models as models

# Transform image in entry
transform = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
])

# Load image and tranform image
img = Image.open("cat.jpg")
img_t = transform(img)
batch_t = torch.unsqueeze(img_t, 0)

# Load models and put it in inference
resnet50 = models.resnet50(pretrained=True)
resnet50.eval()

# Predict the batch of images
out = resnet50(batch_t)
out_softmax = torch.nn.functional.softmax(out, dim=1)[0]

# Sort predictions
_, indices = torch.sort(out, descending=True)

n = 10
# Print top n
top_n = [(idx, out_softmax[idx].item()) for idx in indices[0][:n]]
```

The result should be something like this
```
[(tensor(209), 0.38216879963874817),
 (tensor(208), 0.2528599202632904),
 (tensor(222), 0.05504917725920677),
 (tensor(178), 0.047909315675497055),
 (tensor(207), 0.04158751294016838),
 (tensor(273), 0.01641162671148777),
 (tensor(176), 0.013999605551362038),
 (tensor(173), 0.013278389349579811),
 (tensor(166), 0.009899969212710857),
 (tensor(852), 0.00935537088662386)]
```

Having a look at  the corresp imagenet we can see that class 209 correponds to Labrador retriever. The prediction is correct.

Then let us try to do it with dd, by first getting a traced model:
```
PATH_DD=TODO
PATH_REPLICATE=TODO
cd "${PATH_DD}/tools/torch"
python trace_torchvision.py resnet50 -v
mv resnet50-pretrained.pt  "PATH_REPLICATE/replicate_pytorch_ddlibtorch"
```

Then let create a dd service.
```
curl -X PUT "http://localhost:8080/services/torch_resnet" -d '{
    "description": "image classification service",
    "mllib": "torch",
    "model": {
        "repository": "PATH_REPLICATE/replicate_pytorch_ddlibtorch"
    },
    "parameters": {
        "input": {
            "connector": "image"
        }
    },
    "type": "supervised"
}
'
```
And launch a prediction as shown #611
```
curl -X POST "http://localhost:8080/predict" -d '{
    "service": "torch_resnet",
    "parameters": {
        "input": {
            "width":224,
            "height":224,
            "rgb":true
        },
        "output": {
            "best":3
        }
    },
    "data":["PATH_REPLICATE/replicate_pytorch_ddlibtorch/data/dog.jpg"]
}
'
```

The result obtained is the following:
```
{"status":{"code":200,"msg":"OK"},"head":{"method":"/predict","service":"torch_resnet","time":151.0},"body":{"predictions":[{"classes":[{"prob":1.0,"cat":"818"},{"prob":0.0,"cat":"665"},{"prob":0.0,"cat":"663"},{"prob":0.0,"cat":"669"},{"prob":0.0,"last":true,"cat":"668"}],"uri":"PATH_REPLICATE/data/dog.jpg"}]}}
```

So we have a probability of 1.0 corresponding to "sports car, sport car". That is kind of a weird.
Note that I did not used any normalization whatsoever for pytorch nor DD.
Now, let us try to put normalization on DD using the same number as in #611.

```
curl -X POST "http://localhost:8080/predict" -d '{
    "service": "torch_resnet",
    "parameters": {
        "input": {
            "width":224,
            "height":224,
            "rgb":true,
            "std": 255
        },
        "output": {
            "best":3
        }
    },
    "data":["PATH_REPLICATE/replicate_pytorch_ddlibtorch/data/dog.jpg"]
}
'
```

Here we obtain:
```
{"status":{"code":200,"msg":"OK"},"head":{"method":"/predict","service":"torch_resnet","time":128.0},"body":{"predictions":[{"classes":[{"prob":0.17704856395721436,"cat":"208"},{"prob":0.14640751481056214,"cat":"209"},{"prob":0.09581456333398819,"cat":"178"},{"prob":0.08591518551111222,"cat":"999"},{"prob":0.07016026228666306,"last":true,"cat":"434"}],"uri":"PATH_REPLICATE/replicate_pytorch_ddlibtorch/data/dog.jpg"}]}}%
```

The results are not the same as the ones obtained with Pytorch however it seems better (208 is golden retriever).

That's weird we put an std here and the results change a LOT. Then I dug a bit the code and I saw that the images are loaded as for Caffe, meaning the images are representing as array of ints. However Pytorch does not work like this and uses images of floats between 0 and 1. Moreover by checking the following https://pytorch.org/tutorials/advanced/cpp_export.html you see that it's is the same for libtorch api for C++. Unfortunately I don't see any transform that will change the input values of an image from int to float. Then the std flag instead is not a real std flag but a flag that allows you to transform an image of int (0 to 255) to float (0 to 1). 
That is why when I put 255.0 the results seem better even if they are not the same as for Pytorch. This should be set implicitly and a real flag std should be indicated. (https://github.com/jolibrain/deepdetect/blob/master/src/backends/torch/torchinputconns.h#L108).

To resume a bit:
```
I was not and am still not able to replicate results obtained with Pytorch and libtorch backend in DD.
- One main cause is that the images are not composed of array of float between 0 and 1; (the std flag in #611 is not used as a real flag std just a way to scale the image between 0 and 1; the scaling could be implicit (because traced model except array of float between 0 and 1, toTensor() scales images from 0,255 to 0,1. In https://github.com/jolibrain/deepdetect/blob/master/src/backends/torch/torchinputconns.h#L106 tensors have value between 0 and 255..) and a real flag std should be used as well;
- Even with this known, the results are still different.
```

**Comment by fantes** [comment]
Hi
thanks for the very detailed report. 

two points : 
- this use of std definitely is confusing, we will see how to make things much more easier to use
- transformations are not the same in your pytorch example and in DD call, could you change
```
transform = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
])
```
to
```
transform = transforms.Compose([
    transforms.Resize(224),
    transforms.ToTensor(),
])
```
(which is what should be done internally by DD) and tell us if you get category 208 or category 209 ? 


thank you a lot

**Comment by YaYaB** [comment]
Thanks for the quick reply:
- Yep It would be nice ^^
- I get the category 208 however the probabilities are very different
```
[(tensor(208), 0.3865862488746643),
 (tensor(209), 0.14043162763118744),
 (tensor(222), 0.04077047482132912),
 (tensor(178), 0.033996108919382095),
 (tensor(207), 0.02579624392092228),
 (tensor(168), 0.022857550531625748),
 (tensor(852), 0.018250342458486557),
 (tensor(166), 0.01778768189251423),
 (tensor(999), 0.016320155933499336),
 (tensor(273), 0.015147094614803791)]
```
0.39 vs 0.18
```
{"status":{"code":200,"msg":"OK"},"head":{"method":"/predict","service":"torch_resnet","time":128.0},"body":{"predictions":[{"classes":[{"prob":0.17704856395721436,"cat":"208"},{"prob":0.14640751481056214,"cat":"209"},{"prob":0.09581456333398819,"cat":"178"},{"prob":0.08591518551111222,"cat":"999"},{"prob":0.07016026228666306,"last":true,"cat":"434"}],"uri":"PATH_REPLICATE/replicate_pytorch_ddlibtorch/data/dog.jpg"}]}}%
```

**Comment by beniz** [comment]
I've tried several things and there are several elements at play:
- std and mean need to be applied after normalization. 

Can you please try modifying `torchinputconns.h` with the following and report on results ?

```c++
if (_std != 1.0)                                                                              
     imgt = imgt.mul(1. / _std);                                                                 
imgt[0][0] = imgt[0][0].sub_(0.485).div_(0.229);                                              
imgt[0][1] = imgt[0][1].sub_(0.456).div_(0.224);                                              
imgt[0][2] = imgt[0][2].sub_(0.406).div_(0.225);
```

- I've tested torchvision 0.4.0 with pytorch 1.3 models in python and they cannot be loaded up with DD + pytorch 1.1 as we are using for now. Traced models don't have the same size with different versions, probably due to tensor changes, maybe not weights.

All current changes to torch inference with DD are in #661

**Comment by YaYaB** [comment]
I can try all of this. However with Pytorch I am not doing any std and mean (only normalization thanks to the transform ToTensor). I am using torch 1.1.0 with torchvision 0.4.1.

**Comment by beniz** [comment]
See https://github.com/jolibrain/deepdetect/pull/661 for addition of `scale` and refactor of `std`, as long as proper calls.

Results are within the distribution of the models as much as I can see. You won't be able to obtain the exact same confidence values from Python and C++ due to many variations of image input processing that modifies the pixel values, and the fact that these models appear very much overfitted with low robustness to slight pixel changes. The `top5` metrics should remain very much the same though.

Thanks for pointing this out and all the details, I'm closing this issue for now as solved by #611, and you can reopen if more trouble remains.
